discovery/scaleway/instance.go

```diff
@@ -0,0 +1,102 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scaleway
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/common/model"
+	"github.com/scaleway/scaleway-sdk-go/api/instance/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+const (
+	instanceLabelPrefix = metaLabelPrefix + "instance_"
+
+	instanceNameLabel           = instanceLabelPrefix + "name"
+	instanceImageNameLabel      = instanceLabelPrefix + "image_name"
+	instanceStateLabel          = instanceLabelPrefix + "status"
+	instanceIDLabel             = instanceLabelPrefix + "id"
+	instanceZoneLabel           = instanceLabelPrefix + "zone"
+	instanceCommercialTypeLabel = instanceLabelPrefix + "commercial_type"
+	instanceTagsLabel           = instanceLabelPrefix + "tags"
+	instancePrivateIPLabel      = instanceLabelPrefix + "private_ip"
+	instancePublicIPLabel       = instanceLabelPrefix + "public_ip"
+	instanceIPv6Label           = instanceLabelPrefix + "ipv6"
+)
+
+func (d *Discovery) listInstanceServers(ctx context.Context) ([]model.LabelSet, error) {
+	api := instance.NewAPI(d.client)
+	zone, _ := d.client.GetDefaultZone()
+	project, _ := d.client.GetDefaultProjectID()
+
+	req := &instance.ListServersRequest{
+		Zone:    zone,
+		Project: scw.StringPtr(project),
+	}
+
+	if d.name != "" {
+		req.Name = scw.StringPtr(d.name)
+	}
+
+	if d.tags != nil {
+		req.Tags = d.tags
+	}
+
+	servers, err := api.ListServers(req, scw.WithAllPages(), scw.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	var targets []model.LabelSet
+	for _, server := range servers.Servers {
+		labels := model.LabelSet{
+			instanceIDLabel:             model.LabelValue(server.ID),
+			instanceNameLabel:           model.LabelValue(server.Name),
+			instanceImageNameLabel:      model.LabelValue(server.Image.Name),
+			instanceZoneLabel:           model.LabelValue(server.Zone.String()),
+			instanceCommercialTypeLabel: model.LabelValue(server.CommercialType),
+			instanceStateLabel:          model.LabelValue(server.State),
+		}
+
+		if len(server.Tags) > 0 {
+			// We surround the separated list with the separator as well. This way regular expressions
+			// in relabeling rules don't have to consider tag positions.
+			tags := separator + strings.Join(server.Tags, separator) + separator
+			labels[instanceTagsLabel] = model.LabelValue(tags)
+		}
+
+		if server.PrivateIP != nil {
+			labels[instancePrivateIPLabel] = model.LabelValue(*server.PrivateIP)
+		}
+
+		if server.PublicIP != nil {
+			labels[instancePublicIPLabel] = model.LabelValue(server.PublicIP.Address.String())
+		}
+
+		if server.IPv6 != nil {
+			labels[instanceIPv6Label] = model.LabelValue(server.IPv6.Address.String())
+		}
+
+		addr := net.JoinHostPort(server.PublicIP.Address.String(), strconv.FormatUint(uint64(d.port), 10))
```

Comment:
are the servers guaranteed to have a public ip address or could this panic?

Reply:
I've added a check on whether or not the `PublicIP` is defined. By doing so, I clone the label set and define a new target in case this IP is defined.